### PR TITLE
Add public IP option to AzureML Computes

### PR DESCRIPTION
**Pull request by Lucasjuv**
Trying to address issue #20078.

I've added the enable_node_public_ip option to AzureML computes to be able to provision secure environments.

### CODE REVIEW

File: internal/services/machinelearning/machine_learning_compute_cluster_resource.go [hunk 1]
```diff
@@ -99,6 +99,14 @@ func resourceComputeCluster() *pluginsdk.Resource {
 				ForceNew: true,
 			},
 
+			"enable_node_public_ip": {
```
Comment (katbyte) [inline internal/services/machinelearning/machine_learning_compute_cluster_resource.go:102]: we typically put enabled on the end of the property name, so for consistentancy with the rest of the provider: 
```suggestion
			"node_public_ip_enabled": {
```

File: internal/services/machinelearning/machine_learning_compute_cluster_resource.go [hunk 2]
```diff
@@ -180,6 +188,7 @@ func resourceComputeClusterCreate(d *pluginsdk.ResourceData, meta interface{}) e
 		VMPriority:             &vmPriority,
 		ScaleSettings:          expandScaleSettings(d.Get("scale_settings").([]interface{})),
 		UserAccountCredentials: expandUserAccountCredentials(d.Get("ssh").([]interface{})),
+		EnableNodePublicIP:     utils.Bool(!d.Get("local_auth_enabled").(bool)),
```
Comment (katbyte) [inline internal/services/machinelearning/machine_learning_compute_cluster_resource.go:191]: you can use the new pointer functions here
```suggestion
		EnableNodePublicIP:     pointer.To(!d.Get("local_auth_enabled").(bool)),
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @Lucasjuv - I think we'll need to read this back into state in the read function? aside from that i've left to comments inline and once address i think this will be good to test and merge

File: internal/services/machinelearning/machine_learning_compute_cluster_resource.go (hunk 1 above)
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource.go]: Done.

File: internal/services/machinelearning/machine_learning_compute_cluster_resource.go (hunk 2 above)
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource.go]: Done.

[the remark this answers is not in the document]
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource.go]: Yes, indeed fixed.

File: internal/services/machinelearning/machine_learning_compute_cluster_resource_test.go [hunk 3]
```diff
@@ -205,6 +205,7 @@ resource "azurerm_machine_learning_compute_cluster" "test" {
   vm_size                       = "STANDARD_DS2_V2"
   machine_learning_workspace_id = azurerm_machine_learning_workspace.test.id
   subnet_resource_id            = azurerm_subnet.test.id
+  enable_node_public_ip         = false
```
Comment (stephybun) [inline internal/services/machinelearning/machine_learning_compute_cluster_resource_test.go:208]:
```suggestion
  node_public_ip_enabled         = false
```

File: internal/services/machinelearning/machine_learning_compute_instance_resource_test.go [hunk 4]
```diff
@@ -186,6 +186,7 @@ resource "azurerm_machine_learning_compute_instance" "test" {
   machine_learning_workspace_id = azurerm_machine_learning_workspace.test.id
   virtual_machine_size          = "STANDARD_DS2_V2"
   authorization_type            = "personal"
+  enable_node_public_ip         = false
```
Comment (stephybun) [inline internal/services/machinelearning/machine_learning_compute_instance_resource_test.go:189]:
```suggestion
  node_public_ip_enabled         = false
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Property names need to be updated in the tests

**Comment by Lucasjuv** [comment]
@stephybun I managed to run the acceptance tests locally but they are hanging once the compute instance is created successfully without public the public IP. I'll try to run the test for the cluster.

**Comment by Lucasjuv** [comment]
I've tried testing the cluster and it complains about vCPU quotas. Even though I have nothing else on the subscription. The Compute Instance test still hangs. I've noticed that on some cases the public access on the workspace is disabled. That might be be due to me adding the private endpoint. I've also tried opening the nsg rules to access the workspace and take a look at the computes on the studio.

**Comment by irakhlin** [comment]
@Lucasjuv any progress on this getting merged in? I was about to push up a commit with these same changes but then found this open pull request. I notice you were having issues with your tests not finishing when no public IP is set with a private endpoint, I think the issue is related to not including a private DNS zone. I believe when deploy azure ML workspace on a private endpoint you will need to add a private DNS zone for ML that is linked to the vnet.
https://learn.microsoft.com/en-us/azure/private-link/private-endpoint-dns?view=azureml-api-2
https://learn.microsoft.com/en-us/azure/machine-learning/how-to-configure-private-link?view=azureml-api-2&tabs=cli#create-a-workspace-that-uses-a-private-endpoint

**Comment by irakhlin** [comment]
@Lucasjuv 
Just as a follow up I went and pulled your branch, adding this stanza in your compute test for example
```
resource "azurerm_private_dns_zone" "test" {
  name                = "privatelink.api.ml.azure.us"
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_private_dns_zone_virtual_network_link" "test" {
  name                  = "test-vlink"
  resource_group_name   = azurerm_resource_group.test.name
  private_dns_zone_name = azurerm_private_dns_zone.test.name
  virtual_network_id    = azurerm_virtual_network.test.id
}
```

and adding
```
  private_dns_zone_group {
    name                 = "test"
    private_dns_zone_ids = [azurerm_private_dns_zone.test.id]
  }
```

In the private endpoint produced a passing test. On other thing you note, you will also need to make a change in the azurerm provider instantiation for the test as it will need this setting, 
```
 features {
    resource_group {
      prevent_deletion_if_contains_resources = false
    }
```
This is done in many other tests and is a result of the application insights leaving behind generated resources and preventing deletion of the resource group.

**Comment by Lucasjuv** [comment]
> @Lucasjuv Just as a follow up I went and pulled your branch, adding this stanza in your compute test for example
> 
> ```
> resource "azurerm_private_dns_zone" "test" {
>   name                = "privatelink.api.ml.azure.us"
>   resource_group_name = azurerm_resource_group.test.name
> }
> 
> resource "azurerm_private_dns_zone_virtual_network_link" "test" {
>   name                  = "test-vlink"
>   resource_group_name   = azurerm_resource_group.test.name
>   private_dns_zone_name = azurerm_private_dns_zone.test.name
>   virtual_network_id    = azurerm_virtual_network.test.id
> }
> ```
> 
> and adding
> 
> ```
>   private_dns_zone_group {
>     name                 = "test"
>     private_dns_zone_ids = [azurerm_private_dns_zone.test.id]
>   }
> ```
> 
> In the private endpoint produced a passing test. On other thing you note, you will also need to make a change in the azurerm provider instantiation for the test as it will need this setting,
> 
> ```
>  features {
>     resource_group {
>       prevent_deletion_if_contains_resources = false
>     }
> ```
> 
> This is done in many other tests and is a result of the application insights leaving behind generated resources and preventing deletion of the resource group.

Thanks I had forgotten about the DNS zone configuration. I was in a hurry with my job But I have added the changes now. I`ll run the tests on my end too. 

Thank you.

**Comment by Lucasjuv** [comment]
> @Lucasjuv I updated the properties in the tests hoping that'd be enough to get this in to this week's release but unfortunately there are still test failures:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccComputeInstance_complete
> === PAUSE TestAccComputeInstance_complete
> === CONT  TestAccComputeInstance_complete
> testcase.go:110: Step 1/2 error: Error running apply: exit status 1
> Error: creating Machine Learning Compute ("Compute (Subscription: \"*******\"\nResource Group Name: \"acctestRG-ml-230419131828695174\"\nWorkspace Name: \"acctest-MLW2304191318286951\"\nCompute Name: \"acctest23041974\")"): machinelearningcomputes.MachineLearningComputesClient#ComputeCreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="UserError" Message="Provided EnableNodePublicIp 'False' requires workspace private endpoint." Details=[]
> with azurerm_machine_learning_compute_instance.test,
> on terraform_plugin_test.tf line 106, in resource "azurerm_machine_learning_compute_instance" "test":
> 106: resource "azurerm_machine_learning_compute_instance" "test" {
> --- FAIL: TestAccComputeInstance_complete (365.64s)
> FAIL
> ```

I was able to run most acceptance tests with the current state:
=== RUN   TestAccComputeInstance_complete
=== PAUSE TestAccComputeInstance_complete
=== CONT  TestAccComputeInstance_complete
--- PASS: TestAccComputeInstance_complete (1360.37s)

I`m working with support to solve an issue with cluster quotas but I believe it should be working now.

**Comment by irakhlin** [comment]
@stephybun now that tests are passing what would be needed to have this merged in for the next release?

**Comment by Lucasjuv** [comment]
Related PR: https://github.com/hashicorp/terraform-provider-azurerm/pull/21706

File: internal/services/machinelearning/machine_learning_compute_cluster_resource_test.go (hunk 3 above)
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource_test.go]: Done.

File: internal/services/machinelearning/machine_learning_compute_instance_resource_test.go (hunk 4 above)
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_instance_resource_test.go]: Done.

File: internal/services/machinelearning/machine_learning_compute_instance_resource_test.go [hunk 5]
```diff
@@ -180,12 +180,44 @@ resource "azurerm_subnet_network_security_group_association" "test" {
   network_security_group_id = azurerm_network_security_group.test.id
 }
 
+resource "azurerm_private_dns_zone" "test" {
```
Comment (mbfrahry) [inline internal/services/machinelearning/machine_learning_compute_instance_resource_test.go:183]: With the above comment, we can revert these changes.

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @Lucasjuv, I've taken a pass at this PR and it's got breaking changes the way it is written. I've detailed what is going on below. We'll most likely want to remove the Default value and just have people pass it in when they want to use it.

File: internal/services/machinelearning/machine_learning_compute_instance_resource.go [hunk 6]
```diff
@@ -273,6 +282,7 @@ func resourceComputeInstanceRead(d *pluginsdk.ResourceData, meta interface{}) er
 		d.Set("authorization_type", string(pointer.From(props.Properties.ComputeInstanceAuthorizationType)))
 		d.Set("ssh", flattenComputeSSHSetting(props.Properties.SshSettings))
 		d.Set("assign_to_user", flattenComputePersonalComputeInstanceSetting(props.Properties.PersonalComputeInstanceSettings))
+		d.Set("node_public_ip_enabled", props.Properties.EnableNodePublicIP)
```
[the remark this answers is not in the document]
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_instance_resource.go:285]: > `EnableNodePublicIP` isn't being returned from the api regardless of whether it is set to true or not. We should probably remove it from this resource if that is the case

Well I tested it with the basic scenario and it is as you said. This property isn`t being returned in the current API version. I believe we need to update the API version the provider is using before adding this.

[the remark this answers is not in the document]
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_instance_resource.go:285]: Maybe this PR could help https://github.com/hashicorp/terraform-provider-azurerm/pull/21910

**Comment by Lucasjuv** [comment]
> Hey @Lucasjuv, I've taken a pass at this PR and it's got breaking changes the way it is written. I've detailed what is going on below. We'll most likely want to remove the Default value and just have people pass it in when they want to use it.

I think that the current API version is indeed not returning the property for the compute instance. I didn't test the cluster yet but when we get that API version update it might work. If not I'll have to open a bug on azure's API spec repo.

File: internal/services/machinelearning/machine_learning_compute_cluster_resource.go [hunk 7]
```diff
@@ -100,6 +100,14 @@ func resourceComputeCluster() *pluginsdk.Resource {
 				ForceNew: true,
 			},
 
+			"node_public_ip_enabled": {
+				Type:         pluginsdk.TypeBool,
+				Optional:     true,
+				Default:      true,
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource.go:106]: Hi @mbfrahry ,

Since it defaults to `true` in the swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2023-04-01/machineLearningServices.json#L3534

Is it possible to set `true` into the state when the value isn't returned in the api? Like what I did in this PR: https://github.com/hashicorp/terraform-provider-azurerm/pull/21706/files#diff-0fd2c699ff7d1352db18ad845037d12bfe9df5d528acdbae00977cfd6dc8c579R281

[the remark this answers is not in the document]
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_cluster_resource.go:106]: Let me try this. It seems like a good idea. Thanks!

File: internal/services/machinelearning/machine_learning_compute_instance_resource.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lucasjuv) [reply internal/services/machinelearning/machine_learning_compute_instance_resource.go]: Removed compute instance changes from this PR.

**Comment by stephybun** [comment]
@Lucasjuv unfortunately I see the same behaviour mentioned by @mbfrahry on compute clusters. If the value hasn't previously been set then it isn't returned by the API so the current implementation would be a breaking change for existing clusters.

@ms-henglu the default value defined in the swagger is for version `2023-04-01` so it looks like we should probably upgrade the SDK before adding support for this?

**Comment by ms-henglu** [comment]
Hi @stephybun ,

My bad, I attached the wrong link, this default value is defined in `2022-05-01` as well, this is the api-version azurerm is using.
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2022-05-01/machineLearningServices.json#L3530

**Review by stephybun — APPROVED** [review]
Apologies - it appears I was looking at an outdated version of the PR. I made a minor change to remove the `RequiredWith` and to add validation in the create to check that `subnet_resource_id` is set when `node_public_ip_enabled` is `false` since otherwise `subnet_resource_id` will always be required even when `node_public_ip_enabled` is set to `true`. 

Other than that LGTM 🎋
